numpy/core/src/multiarray/buffer.c

```diff
@@ -596,6 +630,13 @@ _buffer_get_info(PyObject *obj)
             item = PyList_GetItem(item_list, PyList_GET_SIZE(item_list) - 1);
             old_info = (_buffer_info_t*)PyLong_AsVoidPtr(item);
 
+            /*
+             * TODO: In rare cases when an array is both C- and
+             *       F-contiguous, and buffers are exported alternating
+             *       we may always have a cache miss even though the correct
+             *       info had already been exported before (causing a memory
+             *       leak).
+             */
             if (_buffer_info_cmp(info, old_info) == 0) {
```

Comment:
If a new buffer info is always constructed anyway, what's the benefit of having a cache at all? Also, it's a list but I only ever see the last item in the list referenced. Couldn't it be replaced with a single object per array? Then there is no memory growth at all.

Reply:
Yeah, it would be nice to skip part of that sometimes... and would be a small speed enhancement.  But, we have to reuse the old one, so that you do not leak memory and we need to hold on to it to be able to free it again.

The problem is that we can only *free* these when the array is deleted, which is the reason for your memory leak.  This is a fundamental issue around the buffer protocol, which actually does also provide the answer to the issue (by providing a `free` function).  But due to some backcompat issues, we can't just use that...

Times change, and the need for that backcompat probably goes down.  Maybe one more thing that would be nice to try out if we had a major release, I doubt that it would create a big issue in most cases.

Reply:
Sorry, the important part is that, even though nobody likes that, arrays are currently mutable both in shape/strides and dtype... So we can't just stick to a single exported buffer object :(.

Reply:
So to sum up, it isn't a cache in the usual sense but instead a store of all buffer objects.

I see this was 577dbbd844b60958a59e062dae428a34a9c3644f which is 11 years old. If the fix 11 years old, perhaps that `PyArgs_ParseTuple` issue has been fixed in the mean time?

Reply:
No, that issue is still in `PyArg_ParseTuple` with an ugly comment as well there.  It may be that *users* of it are less now though.  And yes, you are right, the name "cache" is a misnomer, since it is persistent.